src/reformat.c

```diff
@@ -220,7 +244,7 @@ avifResult avifImageRGBToYUV(avifImage * image, const avifRGBImage * rgb)
                             a = rgb->pixels[state.rgbOffsetBytesA + (i * state.rgbPixelBytes) + (j * rgb->rowBytes)] / rgbMaxChannelF;
                         }
 
-                        if (!rgb->alphaPremultiplied && image->alphaPremultiplied) {
+                        if (state.toYUVPremultiply) {
```

Comment:
I have a pending pull request that changes the conditional structures of the code here. Please take a look: https://github.com/AOMediaCodec/libavif/pull/510

An observation in that pull request is that we should not enter the outer if statement at line 238 unless we need to premultiply or unpremultiply.